device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+            this._amqp.putToken(SharedAccessSignature.parse(sharedAccessSignature, ['sr', 'sig', 'se']).sr, sharedAccessSignature, (err) => {
+              if (err) {
+                this._amqp.disconnect(() => {
+                  updateSasCallback(getTranslatedError(err, 'AMQP Transport: Could not authorize with puttoken'));
+                });
+              } else {
+                /*Codes_SRS_NODE_DEVICE_AMQP_06_011: [The `updateSharedAccessSignature` method shall call the `done` callback with a null error object and a SharedAccessSignatureUpdated object as a result, indicating the client does NOT need to reestablish the transport connection.]*/
+                updateSasCallback(null, new results.SharedAccessSignatureUpdated(false));
+              }
+            });
+          },
+          getTwinReceiver: (callback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_026: [** The `getTwinReceiver` method shall call the `done` callback with a `null` error argument and the `AmqpTwinClient` instance currently in use.]*/
+            callback(null, this._twinClient);
+          },
+          sendTwinRequest: (method, resource, properties, body, sendTwinRequestCallback) => {
+            this._twinClient.sendTwinRequest(method, resource, properties, body, sendTwinRequestCallback);
+          },
+          attachC2DLink: (callback) => {
+            debug('attaching C2D link');
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_030: [The `Amqp` object shall attach the C2D `ReceiverLink` object if necessary to start receiving messages.]*/
+            this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, receiverLink) => {
+              if (err) {
+                debug('error creating a C2D link: ' + err.toString());
+                handleResult('AMQP Transport: Could not attach link', callback)(err);
+              } else {
+                debug('C2D link created and attached successfully');
+                this._c2dLink = receiverLink;
+                this._c2dLink.on('error', this._c2dErrorListener);
+                this._c2dLink.on('message', this._c2dMessageListener);
+                callback();
+              }
+            });
+          },
+          detachC2DLink: (callback) => {
+            this._stopC2DListener(undefined, callback);
+          },
+          onDeviceMethod: (methodName, methodCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_022: [The `onDeviceMethod` method shall call the `onDeviceMethod` method on the `AmqpDeviceMethodClient` object with the same arguments.]*/
+            this._deviceMethodClient.onDeviceMethod(methodName, methodCallback);
+            this._deviceMethodClient.attach((err) => {
```

Comment:
>attach [](start = 37, length = 6)

silly question: why do we need to call attach here?  Shouldn't the deviceMethodClient object call it's own attach function as a side effect of the onDeviceMethod call?

Reply:
we could. The reason I don't want to is because onDeviceMethod doesn't have a way to surface errors if attaching the link automatically fails (except by using the error event which is a pain to handle). 
Separating the attach calls makes this (and therefore the retry logic) way easier.

---
In reply to: [141133862](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141133862) [](ancestors = 141133862)